kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/stores/AsyncFlushingKeyValueStore.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright 2024 Responsive Computing, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package dev.responsive.kafka.api.async.internals.stores;
+
+import dev.responsive.kafka.api.async.internals.stores.StreamThreadFlushListeners.AsyncFlushListener;
+import java.util.List;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.StateStoreContext;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.internals.CacheFlushListener;
+import org.apache.kafka.streams.state.internals.CachedStateStore;
+import org.apache.kafka.streams.state.internals.CachingKeyValueStore;
+import org.apache.kafka.streams.state.internals.WrappedStateStore;
+import org.slf4j.Logger;
+
+/**
+ * Simple wrapper class around Kafka Streams' {@link CachingKeyValueStore} class that
+ * we use to hook into the {@link CachedStateStore#flushCache()} API and ensure that
+ * all async processing is completed before Streams can proceed with a commit.
+ * <p>
+ * Note: this relies on non-public APIs in Streams and is therefore not protected
+ * by the usual compatibility guarantees provided by Apache Kafka. It may break
+ * upon upgrade. (This is also why it's placed in the o.a.k.streams.state.internals
+ * package -- so we can call the package-private constructor of the super class)
+ */
+public class AsyncFlushingKeyValueStore
```

Comment:
One idea here that I'm not totally convinced is a good idea yet. Instead of wrapping every connected store in this flush listening wrapper, what if we instead just created a single designated no-op store per async processor whose only job is to listen in on flushes. We'd need to:
1. Implement a no-op store. It just needs to implement `CachingStateStore`, do nothing on put/get, and call a listener on `flush`
2. Designate some name for the store. We'd need to figure out how to compute this, but it could just be random or a UUID.
3. Pass the name to the async processor. The async processor looks up the store with this name and sets the listener on it.
4. Async processor would need to re-flush all the stores on its flush listener, since there's no guarantee that the designated store is flushed first. This is the main drawback to this approach IMO.
5. We'd need some way to pass the store to the streams builder. I think we could solve this by requiring the user to call some static method we provide to use the async processor. So instead of `builder.addProcessor(...).addStateStore(...)` you'd do something like `ResponsiveAsyncProcessor.addToTopology(builder, processor, stores)`

The benefits:
- it solves the problem of requiring a store to use the async processor
- simpler connection logic - the async processor just has to look up this store by name

Reply:
I definitely see where you're coming from here, it's just not immediately obvious to me how we can do this. Anyways just giving an update that I'm looking it it next